Source/Microsoft.Teams.Apps.CompanyCommunicator.Prep.Func/PreparingToSend/GetRecipientDataBatches/ProcessRecipientDataListActivity.cs

```diff
@@ -0,0 +1,145 @@
+        /// <param name="notificationDataEntityId">Notification data entity id.</param>
+        /// <returns>Recipient data batches.</returns>
+        public async Task<IEnumerable<IEnumerable<UserDataEntity>>> RunAsync(
+            DurableOrchestrationContext context,
+            string notificationDataEntityId)
+        {
+            var recipientDataBatches =
+                await context.CallActivityWithRetryAsync<IEnumerable<IEnumerable<UserDataEntity>>>(
+                    nameof(ProcessRecipientDataListActivity.ProcessRecipientDataListAsync),
+                    new RetryOptions(TimeSpan.FromSeconds(5), 3),
+                    notificationDataEntityId);
+
+            return recipientDataBatches;
+        }
+
+        /// <summary>
+        /// This method represents the "process recipient data list" activity.
+        /// It processes incoming "recipient data list" as follows.
+        /// 1). Load sent notification data.
+        /// 2). Update total recipient count in notification data entity.
+        /// 3). Page the recipient data list.
+        /// </summary>
+        /// <param name="notificationDataEntityId">Notification id.</param>
+        /// <returns>A <see cref="Task"/> representing the asynchronous operation.</returns>
+        [FunctionName(nameof(ProcessRecipientDataListAsync))]
+        public async Task<IEnumerable<IEnumerable<UserDataEntity>>> ProcessRecipientDataListAsync(
+            [ActivityTrigger] string notificationDataEntityId)
+        {
+            var sentNotificationDataEntityList =
+                await this.sentNotificationDataRepository.GetAllAsync(notificationDataEntityId);
+            var recipientDataList = sentNotificationDataEntityList
+                .Where(p => p.StatusCode == 0)
+                .Select(p =>
+                    new UserDataEntity
+                    {
+                        AadId = p.AadId,
+                        UserId = p.UserId,
+                        ConversationId = p.ConversationId,
+                        ServiceUrl = p.ServiceUrl,
+                        TenantId = p.TenantId,
+                    });
+
+            await this.SetTotalRecipientCountInNotificationDataAsync(
+                notificationDataEntityId,
+                recipientDataList);
+
+            var paged = this.CreateRecipientDataBatches(recipientDataList.ToList());
+            return paged;
+        }
+
+        /// <summary>
+        /// Set total recipient count in notification data entity.
+        /// </summary>
+        /// <param name="notificationDataEntityId">Notification data entity id.</param>
+        /// <param name="recipientDataList">Recipient data list.</param>
+        /// <returns>A <see cref="Task"/> representing the result of the asynchronous operation.</returns>
+        internal async Task SetTotalRecipientCountInNotificationDataAsync(
+            string notificationDataEntityId,
+            IEnumerable<UserDataEntity> recipientDataList)
```

Comment:
Change this to IList, since you turn the enumerable to a list anyway on line 88.
Then change the Count() on line 107 to .Count (property instead of method).